contracts/actions/mcd/McdFLRepayComposite.sol

```diff
@@ -0,0 +1,195 @@
+            tokens,
+            amounts,
+            abi.encode(_repayParams, address(this))
+        );
+        IManager(MCD_MANAGER_ADDR).cdpAllow(_repayParams.vaultId, ACTION_ADDR, 0);
+    }
+
+    /// @notice Balancer FL callback function that formats and calls back this action contract
+    function receiveFlashLoan(
+        address[] memory _tokens,
+        uint256[] memory _amounts,
+        uint256[] memory _feeAmounts,
+        bytes memory _userData
+    ) external override nonReentrant {
+        require(msg.sender == VAULT_ADDR, "Untrusted lender");
+        (RepayParams memory repayParams, address proxy) = abi.decode(_userData, (RepayParams, address));
+
+        _repay(proxy, repayParams);
+
+        uint256 flPaybackAmount = _amounts[0] + _feeAmounts[0];
+        _tokens[0].withdrawTokens(address(VAULT_ADDR), flPaybackAmount);
+    }
+
+    /// @notice Executes flashloan repay logic
+    function _repay(address _proxy, RepayParams memory _repayParams) internal {
+        uint256 collateral = getAllColl(IManager(MCD_MANAGER_ADDR), _repayParams.joinAddr, _repayParams.vaultId);
+        uint256 repayAmount = _repayParams.exchangeData.srcAmount > collateral ? collateral : _repayParams.exchangeData.srcAmount;
+        _repayParams.exchangeData.srcAmount = repayAmount;
+
+        // Sell flashloaned collateral asset for debt asset
+        (uint256 exchangedAmount, ) = _dfsSell(_repayParams.exchangeData, address(this), address(this), false);
+
+        // Take gas fee
+        uint256 paybackAmount = _takeFee(GasFeeTakerParams(
+            _repayParams.gasUsed,
+            DAI_ADDR,
+            exchangedAmount,
+            0
+        ));
+
+        // Withdraw collateral and payback debt
+        {
+            // if paybackAmount is higher than current debt, repay all debt and send remaining dai to proxy
+            (address urn, bytes32 ilk) = getUrnAndIlk(MCD_MANAGER_ADDR, _repayParams.vaultId);
+            uint256 debt = getAllDebt(address(vat), urn, urn, ilk);
+            if (paybackAmount > debt) {
+                DAI_ADDR.withdrawTokens(IDSProxy(_proxy).owner(), paybackAmount - debt);
+                paybackAmount = debt;
+            }
+
+            DAI_ADDR.approveToken(DAI_JOIN_ADDR, paybackAmount);
+            IDaiJoin(DAI_JOIN_ADDR).join(urn, paybackAmount);
+            uint256 daiVatBalance = vat.dai(urn);
+            int256 paybackAmountNormalized = normalizePaybackAmount(address(vat), daiVatBalance, urn, ilk);
+            uint256 frobAmount = convertTo18(_repayParams.joinAddr, repayAmount);
+
+            IManager(MCD_MANAGER_ADDR).frob(
+                _repayParams.vaultId,
+                -toPositiveInt(frobAmount),
+                paybackAmountNormalized
+            );
+            IManager(MCD_MANAGER_ADDR).flux(_repayParams.vaultId, address(this), frobAmount);
+            // withdraw the tokens from Join
+            IJoin(_repayParams.joinAddr).exit(address(this), repayAmount);
```

Comment:
If FL has fee this will return amount without fee and will not be able to repay FL